torch/autograd/grad_mode.py

```diff
@@ -31,13 +32,27 @@ def _wrap_generator(self, func):
         @functools.wraps(func)
         def generator_context(*args, **kwargs):
             gen = func(*args, **kwargs)
-            while True:
-                try:
-                    with self.__class__():
-                        x = next(gen)
-                    yield x
-                except StopIteration:
-                    break
+            grad_mode_context = type(self)()
```

Comment:
I don't think this will work with the `set_grad_enabled` context manager as it sets the state on init and not enter (because we use it as a regular function as well...).
I whink you want to move the creation back into the for loop below.

And also add a test for the other context mangers that use this as base class.

Reply:
Thank you for pointing this out! I have completely missed this.

Reply:
@albanD I am afraid i misunderstood you, but `set_grad_enabled` class [does not inherit](https://github.com/ivannz/pytorch/blob/d8ef729ac25dc4bd79b1545256490289da7675fa/torch/autograd/grad_mode.py#L147) from [_DecoratorContextManager](https://github.com/ivannz/pytorch/blob/d8ef729ac25dc4bd79b1545256490289da7675fa/torch/autograd/grad_mode.py#L17) and does not seem to implement decorator functionality otherwise: it only has `__enter__` and `__exit__`, but no `__call__`.

Reply:
Ho sorry I though this one was inheriting from `_DecoratorContextManager` like all the others but it is not.
So you can ignore that specific comment!

But I still think that delaying the creation would be safer in case that class does things during the init.

Reply:
Agreed, I pushed the patch with delayed instantiation.